### PR TITLE
Final part of extension system - configuration

### PR DESCRIPTION
**Pull request by everzet**
I've splitted container into separate clean object and implemented simple config reader with extensions support.

The way it works is very simple - phpspec2 has it's own ServiceContainer. Whole application class tree lives inside this container and extension system just exposes this container to your custom extension class.

All you need to do is to create custom class, implement `PHPSpec2\Extension\ExtensionInterface` with it:

``` php
<?php

namespace Your\Custom;

use PHPSpec2\Extension\ExtensionInterface;
use PHPSpec2\ServiceContainer;

class Extension implements ExtensionInterface
{
    public function initialize(ServiceContainer $c)
    {
        $c->set('format', 'pretty');

        $c->extend('specification_initializers', function($c) {
            return new CustomSpecInitializer();
        });
    }
}
```

and enable it inside `phpspec.yml`:

``` yaml
extensions:
    - Your\Custom\Extension
```

### CODE REVIEW

File: src/PHPSpec2/Console/Command/RunCommand.php [hunk 1]
```diff
@@ -34,22 +34,22 @@ public function __construct()
...
 
         $result = 0;
         $startTime = microtime(true);
         foreach ($specs as $spec) {
-            $result = max($result, $app['runner']->runSpecification($spec));
+            $result = max($result, $c('runner')->runSpecification($spec));
         }
 
-        $app['event_dispatcher']->dispatch('afterSuite', new Event\SuiteEvent(
+        $c('event_dispatcher')->dispatch('afterSuite', new Event\SuiteEvent(
```
Comment (stof) [inline src/PHPSpec2/Console/Command/RunCommand.php:52]: I don't think this way to call the callable works on PHP 5.3. IIRC, it only works for closures.

File: src/PHPSpec2/ServiceContainer.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
+<?php
+
+namespace PHPSpec2;
+
+use Closure;
+use ArrayAccess;
+use InvalidArgumentException;
+
+class ServiceContainer implements ArrayAccess
```
Comment (stof) [inline src/PHPSpec2/ServiceContainer.php:9]: Why reimplementing Pimple instead of adding it as a dependency ?

Comment (stof) [inline src/PHPSpec2/ServiceContainer.php:9]: hmm, actually, the `extend` method has a different behavior than in Pimple. But all others are the same.

Comment (schmittjoh) [inline src/PHPSpec2/ServiceContainer.php:9]: Less dependencies are better for a testing library imo.

File: src/PHPSpec2/Console/Command/RunCommand.php (hunk 1 above)
Comment (everzet) [inline src/PHPSpec2/Console/Command/RunCommand.php:52]: `__invoke` was been added in 5.3. It works with both 5.3 & 5.4

File: src/PHPSpec2/ServiceContainer.php (hunk 2 above)
Comment (everzet) [inline src/PHPSpec2/ServiceContainer.php:9]: :+1: to @schmittjoh

Plus, i never really evaluated Pimple as a standalone library. Single class with 8 methods just could not worth users time during installation. In fact, i've wrote this custom `ServiceContainer` faster than Composer would fetch Pimple on my machine :-D Jokes aside, making single-8methods-class dependencies is dependency management gone mad for me.

**Comment by MarcelloDuarte** [comment]
What is the dic-cleanup branch? Do we require it for the merge?

**Comment by henrikbjorn** [comment]
:+1:

**Pull request merged**